### PR TITLE
[Clang] Support initializing structured bindings from an array with direct-list-initialization

### LINKED ISSUE #31813: Clang doesn't support structured binding of initializer form {" expr }" for array
|  |  |
| --- | --- |
| Bugzilla Link | [32466](https://llvm.org/bz32466) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @AaronBallman,@willwray |

## Extended Description 
Clang doesn't support structured binding of initializer form {" expr }" if expr is id-expression of array object.
 
Clang does not accept following code.

    int expr[] = { 1,2,3 } ;
    auto [a,b,c]{expr} ;

with the error:
prog.cc:4:18: error: cannot initialize an array element of type 'int' with an lvalue of type 'int [3]'

Other cases are works as expected.
"( expr  )" works as well as "{ expr }" where expr is id-expression of class object:
 
auto [a,b,c](expr) ;

struct X { int x,y,z } ;
X x {1,2,3} ;
auto [a,b,c]{ x } ;

### PR DESCRIPTION
**Pull request by zwuis**
When initializing structured bindings from an array with direct-list-initialization, array copy will be performed, which is a special case not following list-initialization.

This PR adds support for this case.

Fixes #31813.

### CODE REVIEW

File: clang/lib/Sema/SemaInit.cpp [hunk 1]
```diff
@@ -4091,6 +4091,16 @@ void InitializationSequence::AddParenthesizedListInitStep(QualType T) {
   Steps.push_back(S);
 }
 
+void InitializationSequence::AddUnwrapInitListAtTheBeginning(
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaInit.cpp:4094]:
```suggestion
void InitializationSequence::AddUnwrapInitListInitStep(
```

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -302,6 +302,8 @@ Bug Fixes to C++ Support
   template depth than the friend function template. (#GH98258)
 - Clang now rebuilds the template parameters of out-of-line declarations and specializations in the context
   of the current instantiation in all cases.
+- Clang now correctly handle initializing array-typed structured bindings with
```
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:305]:
```suggestion
- Clang now correctly handle direct-list-initialization of a structured bindings from an array
```

File: clang/include/clang/Sema/Initialization.h [hunk 3]
```diff
@@ -1384,6 +1384,11 @@ class InitializationSequence {
 
   void AddParenthesizedListInitStep(QualType T);
 
+  /// Only used when initializing array-typed structured bindings with
+  /// direct-list-initialization. Unwrap the initializer list to get the array
+  /// for array copy.
```
Comment (cor3ntin) [inline clang/include/clang/Sema/Initialization.h:1389]:
```suggestion
  /// Only used when initializing structured bindings from an array with
  /// direct-list-initialization. Unwrap the initializer list to get the array
  /// for array copy.
```

File: clang/test/SemaCXX/cxx1z-decomposition.cpp [hunk 4]
```diff
@@ -198,4 +198,33 @@ namespace lambdas {
...
+    int arr[3]{1, 2, 3};
+    auto [a1, b1, c1](arr);
+    explicit_copy ec_arr[2];
+    auto [a2, b2](ec_arr);
+    arr[0]--;
+    return a1 + b1 + c1 + arr[0];
+  }
+  static_assert(direct_initialization_for_elements() == 6);
+
+  void copy_initialization_for_elements() {
```
Comment (cor3ntin) [inline clang/test/SemaCXX/cxx1z-decomposition.cpp:217]: Can you add a test to check that the elements are correctly set to their value? (like you do above)

**Review by cor3ntin — COMMENTED** [review]
Looking pretty good but I had a few suggestions. Thanks!

**Comment by zwuis** [comment]
Suggestions are applied. Thank you for your review!

I don't have commit accees. Please help me merge this PR if it's ready.

**Review by cor3ntin — APPROVED** [review]

**Review by cor3ntin — APPROVED** [review]

**Pull request merged**